Document --init flag for docker run

nathanleclaire opened the pull request:
@londoncalling @mstanleyjones @johndmulhausen PTAL

nathanleclaire committed:
Document --init flag for docker run

Signed-off-by: Nathan LeClaire <nathan.leclaire@gmail.com>

mstanleyjones commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
 
+## Specifying an init process
+
+Using the `--init` or `--init-path` flags you can specify a process which should
```
You can use the `--init`... to specify a process which should be used

mstanleyjones commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
+
+Using the `--init` or `--init-path` flags you can specify a process which should
+be used as the PID1 in the container, and which binary to use for this init
+process, respectively.  This can be used to ensure the usual responsibilties of
```
No double spacing between sentences, please

mstanleyjones replied:
Specifying an init process ensures that the usual ...

nathanleclaire commented:
Feedback addressed!

justincormack commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
+performed inside the created container.
+
+The default init process used when specifying `--init` without `--init-path` is
+[tini](https://github.com/krallin/tini).
```
Thats a bit confusing - the binary is `docker-init` as found in `PATH`, which is `tini`.

I think it is best to say that the path points to the `docker-init` executable. This is not intended for users to replace with another `init`.

nathanleclaire replied:
> This is not intended for users to replace with another init.

So the flag is intended for specifying a different path for `tini` (named as `docker-init`) specifically?

thaJeztah replied:
The reason it was implemented, is that some distros install the `docker-init` binary in a different path than the default.

I actually wonder if there's a real use-case for having `--init-path` on `docker run` / `docker create`; it's already in a release, so I guess there's no real path back, just wondering if it's feature creep

nathanleclaire replied:
> The reason it was implemented, is that some distros install the docker-init binary in a different path than the default.

Why not use `exec.LookPath` then? :) (Maybe I should just leave addressing `--init-path` out since it seems to be somewhat a leaky abstraction?)

thaJeztah replied:
> Why not use `exec.LookPath` then? :)

On those distros, `docker-init`, `docker-containerd`, etc, are deliberately not in `$PATH`, so that they don't auto-complete on the commandline (and are only started through the systemd unit file, using the full path)

Maybe we should leave it out for now, until the exact use-case for setting a different `--init` _per container_ is clear

LK4D4 commented:
ping @mstanleyjones

mstanleyjones commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
 which may mean you can have a more restrictive set of filters.
 For more details, see the [kernel documentation](https://www.kernel.org/doc/Documentation/prctl/no_new_privs.txt).
 
+## Specifying an init process
```
"Specify an `init` process"

mstanleyjones commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
+responsibilties of an init system, such as reaping zombie processes, are
+performed inside the created container.
+
+The default init process used when specifying `--init` without `--init-path` is
```
Maybe:

The default `init` process used when `--init-path` is not specified is the first `docker-init` executable found in the path. To check this, run `which docker-init`. The default `docker-init` binary is provided by [tini](link-to-tini){: target="_blank" class="_" }

(all of that makes the link to tini open in a new window, without wrecking your text editor's notion of underscores meaning italics)

mstanleyjones commented on docs/reference/run.md:
```diff
@@ -667,6 +667,17 @@ It also causes any seccomp filters to be applied later, after privileges have be
+The default init process used when specifying `--init` without `--init-path` is
+[tini](https://github.com/krallin/tini).
+
 ## Specifying custom cgroups
```
Not your but maybe you can change all these to imperative form "Specify custom cgroups"...

nathanleclaire replied:
Sure

nathanleclaire committed:
Remove --init-path mention

Signed-off-by: Nathan LeClaire <nathan.leclaire@gmail.com>

nathanleclaire commented:
OK I think `--init-path` mention is really just too confusing for something that is more an implementation detail than feature (I had assumed it was for alternate inits, but it's actually for cases where `docker-init` is not found in `$PATH`), so I removed it from this PR.  Please take a look at this version that is more concise and has only `--init` mentioned.  I addressed some of the other comments as well.

nathanleclaire commented:
@mstanleyjones WDYT?

vdemeester reviewed (approved):
LGTM 🐸
/cc @thaJeztah

mdlinville reviewed (approved):
LGTM
